File: new/cpp/3_longest_substring_without_repeating_characters_brute_force.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        // idea: brute force
        // for every character in the given string, iterate starting from the next character until the end
        // and find longest substring without repeating characters. 
        // use unordered_set to tract non-duplicate characters. 
        // if duplicate characters found, break and move to the next character. 

        int max_window_size = 0;

        for (int i = 0 ; i < s.size(); i++) {
            unordered_set<char> window;
            for (int j = i; j < s.size(); j++) {
                if (window.find(s[j]) != window.end()) {
                    // here set.find(something) == set.end(), it means it was not in the given set
                    break;
                }
                window.insert(s[j]);
            }

            max_window_size = max(max_window_size, (int)window.size());
            // window.size() return size_t, need to convert it to int to use with max otherwise throws error
        }

        return max_window_size;
    }
    // time: O(m * n)
    // space: O(m)
};

File: new/python/3_longest_substring_without_repeating_characters_brute_force.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        # idea: brute force
        # have a max so far variable to track the longest valid substring.
        # iterate through the given string, for each character (starting point)
        # have a set to have list of seen characters from that starting point, if at any point, we see a duplicate character,
        # stop the loop, and update the max so far variable.
        max_so_far = 0
        for starting_index in range(len(s)):
            unique_set = set()
            for ending_index in range(starting_index, len(s)):
                if s[ending_index] in unique_set:
                    break
                else:
                    unique_set.add(s[ending_index])
            max_so_far = max(max_so_far, len(unique_set))

        return max_so_far
        # time: O(n * m)
        # space: O(m)
        # n - size of the given string
        # m - size of the longest substring that we find in the inner loop
